From e127d2945798f6e8d3916c16c29a078a600be0e7 Mon Sep 17 00:00:00 2001
From: Myron Stowe <myron.stowe@redhat.com>
Date: Fri, 5 Sep 2014 02:16:08 -0400
Subject: [pci] Add function 1 DMA alias quirk for Marvell devices

Message-id: <20140905021608.32574.51051.stgit@gir.stowe>
Patchwork-id: 91364
O-Subject: [RHEL7.1 PATCH 05/17] PCI: Add function 1 DMA alias quirk for Marvell devices
Bugzilla: 1136537
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Stefan Assmann <sassmann@redhat.com>
RH-Acked-by: Alex Williamson <alex.williamson@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1136537
Upstream Status: cc346a4714a59d08c118e8f33fd86692d3563133

Signed-off-by: Jarod Wilson <jarod@redhat.com>

Conflicts:
  ./drivers/pci/quirks.c
    Hunk #1; the reference to 'dma_alias_devfn' had to be adjusted due
    to KABI (see RHEL7's back-port related to upstream commit 31c2b81).

commit cc346a4714a59d08c118e8f33fd86692d3563133
Author: Alex Williamson <alex.williamson@redhat.com>
Date:   Wed May 28 14:54:00 2014 -0600

    PCI: Add function 1 DMA alias quirk for Marvell devices

    Several Marvell devices and a JMicron device have a similar DMA requester
    ID problem to Ricoh, except they use function 1 as the PCIe requester ID.
    Add a quirk for these to populate the DMA alias with the correct devfn.

    Link: https://bugzilla.kernel.org/show_bug.cgi?id=42679
    Tested-by: George Spelvin <linux@horizon.com>
    Tested-by: Andreas Schragle <ajs124.ajs124@gmail.com>
    Tested-by: Tobias N <qemu@suppser.de>
    Tested-by: <daxcore@online.de>
    Signed-off-by: Alex Williamson <alex.williamson@redhat.com>
    Signed-off-by: Bjorn Helgaas <bhelgaas@google.com>

diff --git a/drivers/pci/quirks.c b/drivers/pci/quirks.c
index 1508802..781111d 100644
--- a/drivers/pci/quirks.c
+++ b/drivers/pci/quirks.c
@@ -3354,6 +3354,43 @@ static void quirk_dma_func0_alias(struct pci_dev *dev)
 DECLARE_PCI_FIXUP_HEADER(PCI_VENDOR_ID_RICOH, 0xe832, quirk_dma_func0_alias);
 DECLARE_PCI_FIXUP_HEADER(PCI_VENDOR_ID_RICOH, 0xe476, quirk_dma_func0_alias);
 
+static void quirk_dma_func1_alias(struct pci_dev *dev)
+{
+ if (PCI_FUNC(dev->devfn) != 1) {
+  dev->pci_dev_rh->dma_alias_devfn =
+   PCI_DEVFN(PCI_SLOT(dev->devfn), 1);
+  dev->dev_flags |= PCI_DEV_FLAGS_DMA_ALIAS_DEVFN;
+ }
+}
+
+/*
+ * Marvell 88SE9123 uses function 1 as the requester ID for DMA.  In some
+ * SKUs function 1 is present and is a legacy IDE controller, in other
+ * SKUs this function is not present, making this a ghost requester.
+ * https://bugzilla.kernel.org/show_bug.cgi?id=42679
+ */
+DECLARE_PCI_FIXUP_HEADER(PCI_VENDOR_ID_MARVELL_EXT, 0x9123,
+    quirk_dma_func1_alias);
+/* https://bugzilla.kernel.org/show_bug.cgi?id=42679#c14 */
+DECLARE_PCI_FIXUP_HEADER(PCI_VENDOR_ID_MARVELL_EXT, 0x9130,
+    quirk_dma_func1_alias);
+/* https://bugzilla.kernel.org/show_bug.cgi?id=42679#c47 + c57 */
+DECLARE_PCI_FIXUP_HEADER(PCI_VENDOR_ID_MARVELL_EXT, 0x9172,
+    quirk_dma_func1_alias);
+/* https://bugzilla.kernel.org/show_bug.cgi?id=42679#c59 */
+DECLARE_PCI_FIXUP_HEADER(PCI_VENDOR_ID_MARVELL_EXT, 0x917a,
+    quirk_dma_func1_alias);
+/* https://bugzilla.kernel.org/show_bug.cgi?id=42679#c46 */
+DECLARE_PCI_FIXUP_HEADER(PCI_VENDOR_ID_MARVELL_EXT, 0x91a0,
+    quirk_dma_func1_alias);
+/* https://bugzilla.kernel.org/show_bug.cgi?id=42679#c49 */
+DECLARE_PCI_FIXUP_HEADER(PCI_VENDOR_ID_MARVELL_EXT, 0x9230,
+    quirk_dma_func1_alias);
+/* https://bugs.gentoo.org/show_bug.cgi?id=497630 */
+DECLARE_PCI_FIXUP_HEADER(PCI_VENDOR_ID_JMICRON,
+    PCI_DEVICE_ID_JMICRON_JMB388_ESD,
+    quirk_dma_func1_alias);
+
 static struct pci_dev *pci_func_0_dma_source(struct pci_dev *dev)
 {
  if (!PCI_FUNC(dev->devfn))
-- 
1.7.1